### ISSUE TITLE
Can not communicate from slave to master

### ISSUE THREAD

**Issue by r-chris** [firehol/netdata#3340]
I am failing to establish a connection between headless netdata slaves and a central host. I have tested this locally and was able to make it work in our local network, but after deployment to the cloud, no slave can get a valid response from the master.

This is the output from one slave (please note that I replaced the URLs)
```
2018-01-25 14:55:34: netdata INFO : STREAM app-staging [send to netdata-master.io:80]: connecting...
2018-01-25 14:55:34: netdata INFO : STREAM app-staging [send to netdata-master.io:80]: initializing communication...
2018-01-25 14:55:34: netdata INFO : STREAM app-staging [send to netdata-master.io:80]: waiting response from remote netdata...
2018-01-25 14:55:34: netdata ERROR: STREAM app-staging [send to netdata-master.io:80]: server is not replying properly (is it a netdata?).
```
On the master I see nothing related to `STREAM` in the `error.log`. As you can see, I am trying to reach the master through port 80. The reason is that I have to deploy my apps as docker containers to a cloud service (cloud foundry), which only allows to expose a single port and maps that (itself) to port 80. I have limited to no control over this. Inside the docker container I run `netdata` and when I test this setup locally, I can in fact get two docker containers (slave & master) to talk to each other.

I read in one of the other issues (https://github.com/firehol/netdata/issues/2153):

> You can stream metrics from netdata to netdata, but neither this protocol is HTTP. You will need to open port 19999.

I assume technically one could use any port (including port 80), but what probably breaks me is the cloud provider putting some web server in front of my instance and forwarding everything through that. In that case I won't be able to add any configuration to that (as explained here: https://github.com/firehol/netdata/wiki/Running-behind-apache).

Do you have any idea what I could do in this situation?

**Comment by r-chris** [comment]
This is my "master" `stream.conf` (I replaced the uuid with a new random one):

```
[stream]
    enabled = no

[C89514EB-04AC-41CD-BD22-F988b94C8AC6]
    enabled = yes
    default history = 3600
    default memory mode = save
    health enabled by default = auto
    allow from = *
```

**Comment by r-chris** [comment]
This is the "slave" `stream.conf`:

```
[stream]
    enabled = yes
    destination = netdata-master.io:80
    api key = C89514EB-04AC-41CD-BD22-F988b94C8AC6
```

**Comment by r-chris** [comment]
Ok, after reading the default https://github.com/firehol/netdata/blob/master/conf.d/stream.conf part about **`This communication is not HTTP (it cannot be proxied by web proxies).`** I assume that that's what is happening here.

Unless anyone has a solution for me, I guess I have to start looking at alternatives.

```
    # Where is the receiving netdata?
    # A space separated list of:
    #
    #      [PROTOCOL:]HOST[%INTERFACE][:PORT]
    #
    # If many are given, the first available will get the metrics.
    #
    # PROTOCOL  = tcp, udp, or unix (only tcp and unix are supported by masters)
    # HOST      = an IPv4, IPv6 IP, or a hostname, or a unix domain socket path.
    #             IPv6 IPs should be given with brackets [ip:address]
    # INTERFACE = the network interface to use (only for IPv6)
    # PORT      = the port number or service name (/etc/services)
    #
    # This communication is not HTTP (it cannot be proxied by web proxies).
    destination =
```

**Comment by ktsaou** [comment]
Hi,

unfortunately the netdata slaves need to connect directly to the master. The port doesn't matter, any will do, but http proxies are not supported.

Your setup will work, if the master can be accessed directly by the slaves. This communication is unidirectional: from all slaves to the master. So, only the master needs to be accessed without a proxy.

Keep in mind netdata supports access lists, so you can configure netdata to listen for STREAM requests to port 19999, but not accept any other requests.

**Comment by r-chris** [comment]
Ok, thanks. I have experimented with the access lists, but in the end I still couldn't expose a second port and have to serve my metrics through some form of http/https.

I started reading about the prometheus interface and decided to use that instead of bothering with any master / slave + netdata registry setup. The [wiki articles](https://github.com/firehol/netdata/wiki/Netdata,-Prometheus,-and-Grafana-Stack) were a great help here! I am now simply exposing the prometheus API and scraping and then tracking that in grafana, which isn't as cool as using the netdata web interface, but it will get the job done.
